Bottom: bd94596808bc8239fa12c660fe278777d193e85e
Top:    41f91d96e5d57ac7f858ecf93a1a835c32a3cabb
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-07 00:31:02 +0100

GTK HID TEST


---

diff --git a/src/hid/gtk/gtkhid-gdk.c b/src/hid/gtk/gtkhid-gdk.c
index e72aa65..73c1463 100644
--- a/src/hid/gtk/gtkhid-gdk.c
+++ b/src/hid/gtk/gtkhid-gdk.c
@@ -709,6 +709,9 @@ ghid_invalidate_lr (int left, int right, int top, int bottom)
   ghid_invalidate_all ();
 }
 
+
+static int invalidate_depth = 0;
+
 void
 ghid_invalidate_all ()
 {
@@ -771,16 +774,21 @@ ghid_invalidate_all ()
   hid_expose_callback (&ghid_hid, &region, 0);
   ghid_draw_grid ();
 
-  DrawAttached ();
+  if (invalidate_depth != 0 && Crosshair.On)
+    printf ("Oh crap - shouldn't be drawing this, invalidate_depth is %i?\n",
+            invalidate_depth);
+  else
+    DrawAttached ();
   DrawMark ();
 
   ghid_screen_update ();
 }
 
+
 void
 ghid_notify_crosshair_change (bool changes_complete)
 {
-  static int invalidate_depth = 0;
+  //static int invalidate_depth = 0;
 
   /* FIXME: We sometimes get called before the GUI is up */
   if (gport->drawing_area == NULL)
diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 791a608..ea77412 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -443,12 +443,10 @@ ghid_port_button_press_cb (GtkWidget * drawing_area,
   state = (GdkModifierType) (ev->state);
   mk = ghid_modifier_keys_state (&state);
   ghid_show_crosshair (FALSE);
-  notify_crosshair_change (false);
 
   do_mouse_action(ev->button, mk);
 
   ghid_invalidate_all ();
-  notify_crosshair_change (true);
   ghid_window_set_name_label (PCB->Name);
   ghid_set_status_line_label ();
   ghid_show_crosshair (TRUE);
@@ -469,13 +467,12 @@ ghid_port_button_release_cb (GtkWidget * drawing_area,
   state = (GdkModifierType) (ev->state);
   mk = ghid_modifier_keys_state (&state);
 
-  notify_crosshair_change (false);
-
   do_mouse_action(ev->button, mk + M_Release);
 
+  notify_crosshair_change (false);
   AdjustAttachedObjects ();
-  ghid_invalidate_all ();
   notify_crosshair_change (true);
+  ghid_invalidate_all ();
 
   ghid_window_set_name_label (PCB->Name);
   ghid_set_status_line_label ();
diff --git a/src/hid/gtk/gui-top-window.c b/src/hid/gtk/gui-top-window.c
index b2f0643..e21ef18 100644
--- a/src/hid/gtk/gui-top-window.c
+++ b/src/hid/gtk/gui-top-window.c
@@ -652,9 +652,8 @@ ghid_menu_cb (GtkAction * action, gpointer data)
     {
       notify_crosshair_change (false);
       AdjustAttachedObjects ();
-      ghid_invalidate_all ();
       notify_crosshair_change (true);
-      ghid_screen_update ();
+      ghid_invalidate_all ();
       ghid_window_set_name_label (PCB->Name);
       ghid_set_status_line_label ();
 #ifdef FIXME
